library/common/main_interface.h

```diff
@@ -1,12 +1,141 @@
 #pragma once
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
 
 // NOLINT(namespace-envoy)
 
+/**
+ * Handle to an Envoy engine instance. Valid only for the lifetime of the engine and not intended
+ * for any external interpretation or use.
+ */
+typedef int envoy_engine_t;
+
+/**
+ * Handle to an outstanding Envoy HTTP request. Valid only for the duration of the request and not
+ * intended for any external interpretation or use.
+ */
+typedef int envoy_request_t;
+
+/**
+ * Handle to an outstanding Envoy HTTP stream. Valid only for the duration of the stream and not
+ * intended for any external interpretation or use.
+ */
+typedef int envoy_stream_t;
+
+/**
+ * Result codes returned by all calls made to this interface.
+ */
+typedef enum { ENVOY_SUCCESS, ENVOY_FAILURE } envoy_status_t;
+
+/**
+ * Error code associated with terminal status of a HTTP stream or request.
+ */
+typedef enum {} envoy_error_code_t;
+
+typedef struct {
+  envoy_status_t status;
+  envoy_request_t request;
+} envoy_request_pair;
+
+typedef struct {
+  envoy_status_t status;
+  envoy_stream_t stream;
+} envoy_stream_pair;
+
+typedef struct {
+  char* name;
+  char* value;
+} envoy_header;
+
+typedef struct {
+  size_t length;
+  envoy_header* headers;
+} envoy_headers;
+
+typedef struct {
+  size_t length;
+  uint8_t* bytes;
+} envoy_data;
+
+const envoy_data envoy_nodata = {0, NULL};
+
+typedef struct {
+  envoy_error_code_t error_code;
+  size_t length;
+  char* string;
+} envoy_error;
+
+#ifdef __cplusplus
+extern "C" { // function pointers
+#endif
+
+typedef void (*on_headers)(envoy_headers headers, bool end_stream);
+typedef void (*on_data)(envoy_data data, bool end_stream);
+typedef void (*on_trailers)(envoy_headers headers);
+typedef void (*on_error)(envoy_error error);
+
+#ifdef __cplusplus
+} // function pointers
+#endif
+
+typedef struct {
+  on_headers h;
```

Comment:
Why not name these more clearly?

Reply:
Definitely open to other naming options. This is technically a callback function that's _part_ of an observer, and not an observer itself (rather, the struct/object that holds it and also contains error handling is the observer). There's also some historical precedent for `on_x` naming of callback functions. It could be called `header_callback_function_ptr`.

Reply:
No strong opinion, that's fine